File: 790. Domino and Tromino Tiling/Solution.java
class Solution {
    long[] results;

    public int numTilings(int n) {
        results = new long[n + 1];
        return (int) numTilingsRec(n);
    }

    public long numTilingsRec(int n) {
        if (n < 0)
            return 0;
        if (n == 0)
            return 1;
        if (n <= 2)
            return n;
        if (n == 3)
            return 5;
        if (results[n] != 0)
            return (int) (results[n] % (Math.pow(10, 9) + 7));
        else {
            results[n] = (2 * numTilingsRec(n - 1)) + numTilingsRec(n - 3);
            return (int) (results[n] % (Math.pow(10, 9) + 7));
        }
    }
}

File: 790. Domino and Tromino Tiling/Solution.py
from typing import List


class Solution:
    def numTilings(self, n: int) -> int:
        return self.numTilingsRecursively(n, [0]*(n+1))

    def numTilingsRecursively(self, n: int, dp: List[int]) -> int:
        if (n < 0):
            return 0
        if (n == 0):
            return 1
        if (n < 3):
            return n
        if (n == 3):
            return 5
        if (not dp[n] == 0):
            return int(dp[n] % (1e9 + 7))
        else:
            result = 2 * \
                self.numTilingsRecursively(
                    n-1, dp)+self.numTilingsRecursively(n-3, dp)
            dp[n] = result
            return int(result % (1e9 + 7))
